pkg/controller/command/verifiable/command.go

```diff
@@ -303,32 +385,207 @@ func (o *Command) GeneratePresentationByID(rw io.Writer, req io.Reader) command.
+	if err != nil {
+		return nil, fmt.Errorf("failed to sign vp: %w", err)
+	}
+
+	return vp, nil
+}
+
+func (o *Command) createAndSignPresentationByID(vc *verifiable.Credential,
+	didDoc *did.Doc) (*verifiable.Presentation, error) {
+	// pk is verification method
+	pk, err := getPublicKeyID(didDoc)
+	if err != nil {
+		return nil, err
+	}
+
+	vp, err := vc.Presentation()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create vp by ID: %w", err)
+	}
+
+	vp, err = o.addLinkedDataProof(vp, pk)
+	if err != nil {
+		return nil, fmt.Errorf("failed to sign vp by ID: %w", err)
+	}
+
+	return vp, nil
+}
+
+func (o *Command) addLinkedDataProof(vp *verifiable.Presentation, pk string) (*verifiable.Presentation, error) {
+	var s signer
+
+	s, err := newKMSSigner(o.ctx.Signer(), o.kResolver, pk)
+	if err != nil {
+		return nil, err
+	}
+
+	signingCtx := &verifiable.LinkedDataProofContext{
+		VerificationMethod:      pk,
+		SignatureRepresentation: verifiable.SignatureJWS,
+		SignatureType:           "Ed25519Signature2018",
+		Suite: ed25519signature2018.New(
+			suite.WithSigner(s)),
+	}
+
+	err = vp.AddLinkedDataProof(signingCtx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to add linked data proof: %w", err)
+	}
+
+	return vp, nil
+}
+
+func (o *Command) parsePresentationRequest(request *PresentationRequest,
+	didDoc *did.Doc) ([]*verifiable.Credential,
+	string, error) {
+	if len(request.VerifiableCredentials) == 0 {
+		return nil, "", fmt.Errorf("no credential found")
+	}
+
+	vcs := make([]*verifiable.Credential, len(request.VerifiableCredentials))
+	// loop through all credentials present in the presentation request
+	for i := range request.VerifiableCredentials {
+		vc, _, err := verifiable.NewCredential([]byte(request.VerifiableCredentials[i]))
+		if err != nil {
+			logutil.LogError(logger, commandName, generatePresentationCommandMethod,
+				"failed to parse presentation request: "+err.Error())
+			return nil, "", fmt.Errorf("parse vc failed: %w", err)
+		}
+
+		vcs[i] = vc
+	}
+
+	verificationMethod, err := getVerificationMethod(request, didDoc)
+	if err != nil {
+		return nil, "", fmt.Errorf("get verification method: %w", err)
+	}
+
+	return vcs, verificationMethod, nil
+}
+func getVerificationMethod(request *PresentationRequest, didDoc *did.Doc) (string, error) {
+	var vm string
+
+	var err error
+
+	if request.ProofOptions.VerificationMethod == "" {
+		// The default implementation if user provides no proofing option, here public key id is verification method
+		vm, err = getPublicKeyID(didDoc)
+		if err != nil {
+			return "", err
+		}
+	} else {
+		vm = request.ProofOptions.VerificationMethod
+	}
+
+	return vm, nil
+}
+
+func getPublicKeyID(didDoc *did.Doc) (string, error) {
+	switch {
+	case len(didDoc.PublicKey) > 0:
+		var publicKeyID string
+
+		for _, k := range didDoc.PublicKey {
+			if strings.HasPrefix(k.Type, Ed25519VerificationKey) {
+				publicKeyID = k.ID
+				break
+			}
+		}
+
+		if !isDID(publicKeyID) {
```

Comment:
In which case public key id is not did? This should not happen.

Reply:
Created todo #1640